datasets/blbooks/blbooks.py

```diff
@@ -0,0 +1,221 @@
+        BritishLibraryBooksConfig(
+            name="1510_1699",
+            description="Subset covering 1510-1699 of " + _DESCRIPTION,
+            data_urls=_1510_1699,
+            citation=_CITATION,
+            url="TODO",
+            skip_empty=True,
+        ),
+    ]
+
+    DEFAULT_CONFIG_NAME = "1500_1899"
+
+    def _info(self):
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=features,
+            supervised_keys=None,
+            homepage="https://www.bl.uk/collection-guides/digitised-printed-books",
+            citation=_CITATION,
+            task_templates=[LanguageModeling(text_column="text")],
+        )
+
+    def _split_generators(self, dl_manager: datasets.DownloadManager):
+        urls_to_download = self.config.data_urls
+        downloaded_archives = dl_manager.download(urls_to_download)
+        downloaded_archives = [dl_manager.iter_archive(archive) for archive in downloaded_archives]
+        return [datasets.SplitGenerator(name=datasets.Split.TRAIN, gen_kwargs={"data_dirs": downloaded_archives})]
+
+    @lru_cache(maxsize=512)
+    def _parse_date(self, date):
+        if date is not None:
+            date = datetime.strptime(str(date), "%Y")
+        return date
+
+    def _parse_data(self, data: Dict) -> Dict:
+        mean_wc_ocr = data["mean_wc_ocr"]
+        mean_wc_ocr = float(mean_wc_ocr) if mean_wc_ocr else None
+        std_wc_ocr = data["std_wc_ocr"]
+        std_wc_ocr = float(data["std_wc_ocr"]) if std_wc_ocr else None
+        date = data["date"]
+        if date is not None:
+            date = datetime.strptime(str(date), "%Y")
+        return {
+            "record_id": data["record_id"],
+            "date": date,
+            "raw_date": data["raw_date"],
+            "title": data["title"],
+            "place": data["place"],
+            "text": data["text"],
+            "pg": int(data["pg"]),
+            "mean_wc_ocr": data["mean_wc_ocr"],
+            "std_wc_ocr": std_wc_ocr,
+            "name": data["Name"],
+            "all_names": data["All names"],
+            "Publisher": data["Publisher"],
+            "Country of publication 1": data["Country of publication 1"],
+            "all Countries of publication": data["All Countries of publication"],
+            "Physical description": data["Physical description"],
+            "Language_1": data["Language_1"],
+            "Language_2": data["Language_2"],
+            "Language_3": data["Language_3"],
+            "Language_4": data["Language_4"],
+            "multi_language": data["multi_language"],
+        }
+
+    def _generate_examples(self, data_dirs):
+        skip_empty = self.config.skip_empty
+        id_ = 0
+        for data_dir in data_dirs:
+            for path, file in data_dir:
+                if not path.endswith(".gz"):
+                    continue
+                with gzip.open(file) as json_l:
+                    for row in json_l.readlines():
```

Comment:
nit: you can iterate one line at a time here
```suggestion
                    for row in json_l:
```